### ISSUE TITLE
Inconsistent behavior for workqueue consumers

### ISSUE THREAD

**Issue by LautaroJayat** [nats-io/nats-server#4957]
### Observed behavior

**TLDR;**
When creating two suscribers with different names and non exclusive filtered subjects sometimes the servers responds with an error and prevents you to create the consumer, but other times it allows creations and prompt you to manually remove it.

**Given the following scenario:**
* One Nats Cluster with Jetstream enabled
* Just one Stream with the following subject: `testing.*`
* Subscriber with name `subscriber-a` suscribes to `testing.>` and everything goes OK.
* Subscriber with name `subscriber-b` tries to join to `testing.>` gets rejected **BUT**:

1. In one case we get this message in the nats console:  `[WRN] Proposing to delete consumer 'APP > suscriber-b > testing.1-v2' due to assignment response error: filtered consumer not unique on workqueue stream (10100)`. The Go client used receives an error and we can handle this easily in the application.

2. In other cases we get this message in the nats console: `[WRN] Consumer "subscriber-b" > "testing.1-v2" has a filter subject that overlaps with other consumers, which is not allowed for a stream with WorkQueue policy, it should be administratively deleted`. The Go client used does not error and the application has no way to know this happened.

### Expected behavior

The cluster should respond with a consistent answer when two subscribers with different names tries to join a stream but they have overlapping subject filters.
If possible, the response should be something like an error, so applications can understand what is happening.

### Server and client version

Client: `github.com/nats-io/nats.go v1.31.0` 

Server: official docker image `nats:latest`

### Host environment

```bash
uname -a
# Linux 6.2.0-39-generic #40~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC x86_64 GNU/Linux
```

### Steps to reproduce

1. Create a nats cluster with Jetstream enabled.
2. Create a stream.
4. Join one go client with the following steps: 

```go
// First client will connect with no problem
nc, _ := nats.Connect( /* nats url */ )
js,  _ := jetstream.New(nc)
consumer, err := js.CreateOrUpdateConsumer(context.TODO(), "testing", jetstream.ConsumerConfig{
	Name:           "testing-v1",
	FilterSubjects: []string{"testing.>"},
})

/***********************************************************************************/

// second client will fail but with no consistent response from server
nc, _ := nats.Connect( /* nats url */ )
js, _ := jetstream.New(nc)
consumer, err := js.CreateOrUpdateConsumer(context.TODO(), "testing", jetstream.ConsumerConfig{
	Name:           "testing-v2",
	FilterSubjects: []string{"testing.>"},
})
```

*Stream configuration*
```json
{
    "config": {
      "name": "testing",
      "subjects": [
        "testing.*"
      ],
      "retention": "workqueue",
      "max_consumers": -1,
      "max_msgs_per_subject": -1,
      "max_msgs": -1,
      "max_bytes": -1,
      "max_age": 0,
      "max_msg_size": -1,
      "storage": "memory",
      "discard": "new",
      "num_replicas": 1,
      "duplicate_window": 120000000000,
      "sealed": false,
      "deny_delete": false,
      "deny_purge": false,
      "allow_rollup_hdrs": false,
      "allow_direct": true,
      "mirror_direct": false
    }}
```

*Cluster configuration*

```conf
# ./testdata/jetstream.conf

jetstream {
    store_dir: /tmp/jetstream/
    max_mem: 1G
    max_file: 100G
}


accounts: {
  $SYS: {
    users: [{user: sys, password: pass}]
  }
  APP: {
    jetstream: true
    users: [{user: user, password: pass}]
  }
}


cluster: {
  name: TESTING,
  port: 6222,
  routes: [
    "nats://na_1:6222"
  ]
}
```

Docker compose for the cluster:

```yaml
version: "3.5"
services:
  na_1:
    image: nats
    command: -c /app/jetstream.conf --name na_1 -p 4222
    volumes:
      - ./testdata:/app
    ports:
      - 4222:4222
    networks: ["nats"]

  na_2:
    image: nats
    command: -c /app/jetstream.conf --name na_2 -p 4223
    volumes:
      - ./testdata:/app
    ports:
      - 4223:4223
    networks: ["nats"]

  na_3:
    image: nats
    command: -c /app/jetstream.conf --name na_3 -p 4224
    volumes:
      - ./testdata:/app
    ports:
      - 4224:4224
    networks: ["nats"]

networks:
  nats:
    name: nats
```

**Comment by derekcollison** [comment]
If the filter subjects are mutually exclusive this should work.

**Comment by derekcollison** [comment]
The subscribers filter subject is wider than the stream, and if another consumer exists that this filter subject overlaps with it will complain about that as well.

**Comment by LautaroJayat** [comment]
But I'm not complaining about that.
What I'm saying that one would expect a consistent error from the server.
Instead, some times we got that error and other times we get a non error response that tells you to remove the consumer by hand.

**Comment by derekcollison** [comment]
Would you happen to have a simple Go program that shows the issue?

**Comment by derekcollison** [comment]
@Jarema any changes we want to make here before next server release?

**Comment by LautaroJayat** [comment]
@Jarema You are the GOAT

**Comment by Jarema** [comment]
@derekcollison I think that the explanation is enough for now. Need to dig deeper to see if changing that behavior is safe.

**Comment by nico-famedly** [comment]
I am hitting a similar issue, with probably the same root cause:

I have one work queue stream to distribute events. I have several consumers, that race to subscribe to any available subject for load balancing reasons on this same stream, but with different filter subjects. Usually this works fine, but if I hit very high event input on the stream, multiple consumers start subscribing to the same subject without one of them getting rejected. This means that under high load, consumers will start doing work twice, which is exactly opposite of what I want. My assumption is, that because of the high event load, the stream starts entering the recovering state for a short time, which then causes duplicate subscriptions.

Judging from the commit, that added this behaviour: https://github.com/nats-io/nats-server/commit/49faba9e33eb702c739c681816475e2a381c4e5a

I think this was added because duplicate subscriptions might already be present and to not interrupt those maybe? The actual fix in that commit seems to just be in stream.go, so I assume the hunk in consumer.go was just added for migration purposes?

**Comment by derekcollison** [comment]
@nico-famedly what server version are you running?

**Comment by nico-famedly** [comment]
@derekcollison 2.10.12

**Comment by derekcollison** [comment]
Are these consumers ephemeral or named? (legacy durable)

**Comment by nico-famedly** [comment]
I think I experimented with a few variations, but

```
                    stream
                        .create_consumer(async_nats::jetstream::consumer::pull::Config {
                            durable_name: Some(format!("myconsumer{}a{}", pid, calculate_hash(&subject))),
                            filter_subject: subject.clone(),
                            inactive_threshold: Duration::from_secs(1),
                            description: Some(format!("Consumer for for {}", subject)),
                            ..Default::default()
                        })
```

was the latest one I experimented with (and where I could reproduce the issue). However I increased the amount of replicas on the topic now, which seems to have "fixed" the issue. I think that is an ephemeral consumer, considering it gets removed automatically after some inactivity, but I am not quite confident in my ability to use those terms yet. I hope the whole (rust) config helps? This is with the rust async-nats 0.34.0 as the client lib.

The stream was created with:

```
    let mut stream = js
        .create_stream(async_nats::jetstream::stream::Config {
            name: "teststream".to_string(),
            subjects: vec!["teststream.>".to_string()],
            retention: jetstream::stream::RetentionPolicy::WorkQueue,
            consumer_limits: Some(jetstream::stream::ConsumerLimits {
                inactive_threshold: Duration::from_secs(1),
                ..Default::default()
            }),
            // num_replicas: 3,
            ..Default::default()
        })
        .await?;
```

And if the replicas were not set to 3 and I had 3 servers in my cluster, something like this:

```
nats pub teststream.r1 --count=100000 --sleep 1ms "publication #{{Count}} @ {{TimeStamp}}"
```

would cause consumers trying to subscribe to `teststream.r1` to both successfully subscribe. However trying to reproduce it now, I actually had to manually kill one of the nats servers to get that behaviour. I don't think I did that originally, but maybe that is required to force a server into the recovering state. I had thought I had managed to do that by just flooding the server with messages though...

I hope this helps! And thank you for the quick responses!

**Comment by derekcollison** [comment]
I figured it out. The issue was that the stream was probably R1 and located on the meta-leader.
We incorrectly marked the consumer as recovering and warned vs not allowing.

PR inbound, and fix will be in 2.10.14 to be released this week.

**Comment by LautaroJayat** [comment]
I apologize for opening this issue and leave it to rot without providing any additional help.
@derekcollison thank you a lot for your work and dedication. Amazing.

**Comment by nico-famedly** [comment]
I have to agree, thanks a lot for the work!
